runtime/autoload/health/nvim.vim

```diff
@@ -0,0 +1,408 @@
+              \ . 'virtualenv.', host_prog_var)
+              \ ])
+      endif
+    endif
+  endif
+
+  if empty(python_bin) && !empty(python_bin_name)
+    " An error message should have already printed.
+    call health#report_error(printf('"%s" was not found.', python_bin_name))
+  elseif !empty(python_bin) && !s:check_bin(python_bin, notes)
+    let python_bin = ''
+  endif
+
+  " Check if $VIRTUAL_ENV is active
+  let virtualenv_inactive = 0
+
+  if exists('$VIRTUAL_ENV')
+    if !empty(pyenv)
+      let pyenv_prefix = resolve(s:trim(system(printf('"%s" prefix', pyenv))))
+      if $VIRTUAL_ENV != pyenv_prefix
+        let virtualenv_inactive = 1
+      endif
+    elseif !empty(python_bin_name) && exepath(python_bin_name) !~# '^'.$VIRTUAL_ENV.'/'
+      let virtualenv_inactive = 1
+    endif
+  endif
+
+  if virtualenv_inactive
+    suggestions = [
+          \ 'If you are using Zsh, see: http://vi.stackexchange.com/a/7654/5229',
+          \ ]
+    call health#report_warn(
+          \ '$VIRTUAL_ENV exists but appears to be inactive. ' 
+          \ . 'This could lead to unexpected results.', 
+          \ suggestions)
+  endif
+
+  " Diagnostic output
+  call health#report_info('Executable:' . (empty(python_bin) ? 'Not found' : python_bin))
+  if len(python_multiple)
+    for path_bin in python_multiple
+      call health#report_info('Other python executable: ' . path_bin)
+    endfor
+  endif
+
+  if !empty(python_bin)
+    let [pyversion, current, latest, status] = s:version_info(python_bin)
+    if a:version != str2nr(pyversion)
+      " call add(notes, 'Warning: Got an unexpected version of Python.  '
+      "       \ . 'This could lead to confusing error messages.  Please '
+      "       \ . 'consider this before reporting bugs to plugin developers.')
+      call health#report_warn('Got an unexpected version of Python.' .
+                  \ ' This could lead to confusing error messages.')
+    endif
+    if a:version == 3 && str2float(pyversion) < 3.3
+      call health#report_warn('Python 3.3+ is recommended.')
+    endif
+
+    call health#report_info('Python Version: ' . pyversion)
+    call health#report_info(printf('%s-neovim Version: %s', python_bin_name, current))
+
+    if current == 'not found'
+      suggestions = [
+            \ 'Use the command `pip' . a:version . 'install neovim`',
+            \ ]
+      call health#report_error(
+            \ 'Neovim Python client is not installed.',
+            \ suggestions)
+    endif
+
+    if latest == 'unknown'
+      call health#report_warn('Unable to fetch latest Neovim Python client version.')
+    endif
+
+    if status == 'outdated'
+      call health#report_warn('Latest Neovim Python client versions: ('.latest.')')
+    else
+      call health#report_ok('Latest Neovim Python client is installed: ('.status.')')
+    endif
+  endif
+
+  call health#report_notes(notes)
+endfunction
+
+
+function! health#nvim#check() abort
+  redir => report
```

Comment:
The `redir` capturing part since that should be the only point of failure for output.  Leaving that up to checker might cause output leaking out of order or an exception thrown without `redir END`